### ISSUE TITLE
Make `AndroidJUnit5JacocoReport` a `@CacheableTask`

### ISSUE THREAD

**Issue by autonomousapps** [mannodermaus/android-junit5#138]
Please note that the superclass, [JacocoReport](https://docs.gradle.org/current/javadoc/org/gradle/testing/jacoco/tasks/JacocoReport.html), is cacheable.

**Why?**
In a test against a project with several hundred unit tests, generating the report took over 5 seconds. Not world-shattering, but definitely a good target for caching (copying the reports from the cache will probably take less time than re-generating them).

**Comment by mannodermaus** [comment]
Good idea. I suppose a unit test for this would look something like this:
- Setup a project
- Run the task
- Run the task again, check outcome for `UP_TO_DATE`?

**Comment by autonomousapps** [comment]
For caching, the outcome would be FROM-CACHE. Assuming, of course, the
build cache has been enable via --build-cache or org.gradle.caching=true.

**Comment by mannodermaus** [comment]
I believe this is currently blocked by [115873047](https://issuetracker.google.com/issues/115873047). The `jacocoTestReport` tasks don't seem to want to be retrieved from the build cache, just like their `test` companions:

**Build 1 (`./gradlew jacocoTestReport --build-cache --stacktrace`)**
(Note: I didn't fully clear the build cache before this)

<details><summary>Output</summary>

```
> Task :preBuild UP-TO-DATE
> Task :preDebugBuild FROM-CACHE
> Task :compileDebugAidl NO-SOURCE
> Task :compileDebugRenderscript NO-SOURCE
> Task :checkDebugManifest
> Task :generateDebugBuildConfig FROM-CACHE
> Task :prepareLintJar UP-TO-DATE
> Task :generateDebugSources UP-TO-DATE
> Task :javaPreCompileDebug FROM-CACHE
> Task :mainApkListPersistenceDebug FROM-CACHE
> Task :generateDebugResValues FROM-CACHE
> Task :generateDebugResources UP-TO-DATE
> Task :mergeDebugResources FROM-CACHE
> Task :createDebugCompatibleScreenManifests FROM-CACHE
> Task :processDebugManifest FROM-CACHE
> Task :processDebugResources FROM-CACHE
> Task :compileDebugJavaWithJavac FROM-CACHE
> Task :generateDebugUnitTestSources UP-TO-DATE
> Task :preDebugUnitTestBuild UP-TO-DATE
> Task :javaPreCompileDebugUnitTest FROM-CACHE
> Task :compileDebugUnitTestJavaWithJavac FROM-CACHE
> Task :processDebugJavaRes NO-SOURCE
> Task :processDebugUnitTestJavaRes NO-SOURCE

> Task :testDebugUnitTest

de.mannodermaus.app.JavaTest > test() PASSED

> Task :jacocoTestReportDebug
> Task :preReleaseBuild FROM-CACHE
> Task :compileReleaseAidl NO-SOURCE
> Task :compileReleaseRenderscript NO-SOURCE
> Task :checkReleaseManifest
> Task :generateReleaseBuildConfig FROM-CACHE
> Task :generateReleaseSources UP-TO-DATE
> Task :javaPreCompileRelease FROM-CACHE
> Task :mainApkListPersistenceRelease FROM-CACHE
> Task :generateReleaseResValues FROM-CACHE
> Task :generateReleaseResources UP-TO-DATE
> Task :mergeReleaseResources FROM-CACHE
> Task :createReleaseCompatibleScreenManifests FROM-CACHE
> Task :processReleaseManifest FROM-CACHE
> Task :processReleaseResources FROM-CACHE
> Task :compileReleaseJavaWithJavac FROM-CACHE
> Task :generateReleaseUnitTestSources UP-TO-DATE
> Task :preReleaseUnitTestBuild UP-TO-DATE
> Task :javaPreCompileReleaseUnitTest FROM-CACHE
> Task :compileReleaseUnitTestJavaWithJavac FROM-CACHE
> Task :processReleaseJavaRes NO-SOURCE
> Task :processReleaseUnitTestJavaRes NO-SOURCE

> Task :testReleaseUnitTest

de.mannodermaus.app.JavaTest > test() PASSED

> Task :jacocoTestReportRelease
> Task :jacocoTestReport

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.2.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 26s
31 actionable tasks: 6 executed, 24 from cache, 1 up-to-date
```

</details>

**Build 2 (`./gradlew clean jacocoTestReport --build-cache --stacktrace`)**

<details><summary>Output</summary>

```
> Task :clean
> Task :preBuild UP-TO-DATE
> Task :preDebugBuild FROM-CACHE
> Task :compileDebugAidl NO-SOURCE
> Task :compileDebugRenderscript NO-SOURCE
> Task :checkDebugManifest
> Task :generateDebugBuildConfig FROM-CACHE
> Task :prepareLintJar UP-TO-DATE
> Task :generateDebugSources UP-TO-DATE
> Task :javaPreCompileDebug FROM-CACHE
> Task :mainApkListPersistenceDebug FROM-CACHE
> Task :generateDebugResValues FROM-CACHE
> Task :generateDebugResources UP-TO-DATE
> Task :mergeDebugResources FROM-CACHE
> Task :createDebugCompatibleScreenManifests FROM-CACHE
> Task :processDebugManifest FROM-CACHE
> Task :processDebugResources FROM-CACHE
> Task :compileDebugJavaWithJavac FROM-CACHE
> Task :generateDebugUnitTestSources UP-TO-DATE
> Task :preDebugUnitTestBuild UP-TO-DATE
> Task :javaPreCompileDebugUnitTest FROM-CACHE
> Task :compileDebugUnitTestJavaWithJavac FROM-CACHE
> Task :processDebugJavaRes NO-SOURCE
> Task :processDebugUnitTestJavaRes NO-SOURCE

> Task :testDebugUnitTest

de.mannodermaus.app.JavaTest > test() PASSED

> Task :jacocoTestReportDebug
> Task :preReleaseBuild FROM-CACHE
> Task :compileReleaseAidl NO-SOURCE
> Task :compileReleaseRenderscript NO-SOURCE
> Task :checkReleaseManifest
> Task :generateReleaseBuildConfig FROM-CACHE
> Task :generateReleaseSources UP-TO-DATE
> Task :javaPreCompileRelease FROM-CACHE
> Task :mainApkListPersistenceRelease FROM-CACHE
> Task :generateReleaseResValues FROM-CACHE
> Task :generateReleaseResources UP-TO-DATE
> Task :mergeReleaseResources FROM-CACHE
> Task :createReleaseCompatibleScreenManifests FROM-CACHE
> Task :processReleaseManifest FROM-CACHE
> Task :processReleaseResources FROM-CACHE
> Task :compileReleaseJavaWithJavac FROM-CACHE
> Task :generateReleaseUnitTestSources UP-TO-DATE
> Task :preReleaseUnitTestBuild UP-TO-DATE
> Task :javaPreCompileReleaseUnitTest FROM-CACHE
> Task :compileReleaseUnitTestJavaWithJavac FROM-CACHE
> Task :processReleaseJavaRes NO-SOURCE
> Task :processReleaseUnitTestJavaRes NO-SOURCE

> Task :testReleaseUnitTest

de.mannodermaus.app.JavaTest > test() PASSED

> Task :jacocoTestReportRelease
> Task :jacocoTestReport

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.2.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 14s
32 actionable tasks: 7 executed, 24 from cache, 1 up-to-date
```

</details>

Adding the annotation to `AndroidJUnit5JacocoReport` is trivial and should land in the next `1.4.1.0-SNAPSHOT`. Not sure if the plugin can do more than that at this point, though.

**Comment by autonomousapps** [comment]
To be clear, the issue with append=true by default has disappeared with Gradle 5. It also _shouldn't_ matter that the test tasks on which the report task depends aren't cacheable. So long as they're determinate (always produce the same outputs), the report task shouldn't care whether those outputs (its inputs) came from the cache or came from a recent execution.

May I ask what build you're testing against? I'd like to run some of my own tests and generate a build comparison with gradle enterprise. It'll show me file-level diffs, which are useful for debugging this kind of thing.

**Comment by mannodermaus** [comment]
I added a barebones project to the plugin's test resources, and executed two consecutive runs through Gradle TestKit. I removed it eventually because it wouldn't say `FROM_CACHE` as expected, but I put it back onto a branch for you to have a look at:

[Functional Test](https://github.com/mannodermaus/android-junit5/blob/feature/item-138/plugin/android-junit5/src/test/kotlin/de/mannodermaus/gradle/plugins/junit5/FunctionalTests.kt#L173-L191)
[Test project root](https://github.com/mannodermaus/android-junit5/tree/feature/item-138/plugin/android-junit5/src/test/projects/jacoco-build-cache)

**Comment by mannodermaus** [comment]
Released in `1.4.1.0`. Let's resume the discussion regarding the build cache another time!

**Comment by autonomousapps** [comment]
Sorry, I've been very busy! I'll take a look at this when I get a chance.
